ipld/retriever.go

```diff
@@ -229,29 +250,45 @@ func (rs *retrievalSession) doRequest(ctx context.Context, q *quadrant) {
 			// the left or the right side of the tree represent some portion of the quadrant
 			// which we put into the rs.square share-by-share by calculating shares' indexes using q.index
 			GetShares(ctx, rs.bget, nd.Links()[q.x].Cid, size, func(j int, share Share) {
-				// the R lock here is *not* to protect rs.square from multiple concurrent shares writes
-				// but to avoid races between share writes and repairing attempts
-				// shares are written atomically in their own slice slot
+				// NOTE: Each share can be fetched twice, from two sources(Row or Col).
+				// These shares are always equal, and we allow only the first one to be written
+				// in the square.
+				// NOTE-2: We never actually fetch shares from the network *twice*,
+				// and they are cached on IPLD(blockservice) level.
```

Comment:
Doesn't Note-2 contradict the first note?
> Each share can be fetched twice, from two sources(Row or Col).
> We never actually fetch shares from the network *twice*

this is confusing.

Reply:
There is a chance that the same share could appear in this func twice, without being fetched twice from the network. Even though we traverse over the same share twice, it's never fetched over the network twice because of the caching on BlockService layer. I really don't know ho to explain this better.

Reply:
I think I understand. And I think you explained it well enough. It just sounds contradictory at first sight but it is actually quite clear.